File: library-core/src/main/java/com/mikepenz/iconics/context/IconicsFactory.java
package com.mikepenz.iconics.context;

import android.annotation.SuppressLint;
import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;

import com.mikepenz.iconics.Iconics;
import com.mikepenz.iconics.IconicsDrawable;
import com.mikepenz.iconics.animation.IconicsAnimatedDrawable;
import com.mikepenz.iconics.core.R;

import androidx.appcompat.view.menu.ActionMenuItemView;

class IconicsFactory {

    View onViewCreated(View view, Context context, AttributeSet attrs) {
        if (view != null && view.getTag(R.id.iconics_tag_id) != Boolean.TRUE) {
            onViewCreatedInternal(view, context, attrs);
            view.setTag(R.id.iconics_tag_id, Boolean.TRUE);
        }
        return view;
    }

    @SuppressLint("RestrictedApi")
    private void onViewCreatedInternal(View view, final Context context, AttributeSet attrs) {
        if (attrs == null) {
            return;
        }

        if (view instanceof ActionMenuItemView) {
            IconicsDrawable drawable = IconicsAttrsApplier.getIconicsDrawable(context, attrs);
            if (drawable != null) {
                ((ActionMenuItemView) view).setIcon(drawable);

                if (drawable instanceof IconicsAnimatedDrawable) {
                    ((IconicsAnimatedDrawable) drawable).animateIn(view);
                }
            }
        } else if (view instanceof EditText) {
            //handle iconics
            new Iconics.IconicsBuilder().ctx(context).on((TextView) view).build();

            //for an editText we only style initial as styling the Editable causes problems!
        } else if (view instanceof TextView) {
            //handle iconics
            new Iconics.IconicsBuilder().ctx(context).on((TextView) view).build();

            ((TextView) view).addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

                }

                @Override
                public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

                }

                @Override
                public void afterTextChanged(Editable editable) {
                    Iconics.styleEditable(context, editable);
                }
            });
        } else if (view instanceof ImageView) {
            IconicsDrawable drawable = IconicsAttrsApplier.getIconicsDrawable(context, attrs);
            if (drawable != null) {
                ((ImageView) view).setImageDrawable(drawable);

                if (drawable instanceof IconicsAnimatedDrawable) {
                    ((IconicsAnimatedDrawable) drawable).animateIn(view);
                }
            }
        }
    }
}

File: library-core/src/main/java/com/mikepenz/iconics/context/IconicsFactory.kt
/*
 * Copyright 2019 Mike Penz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mikepenz.iconics.context

import android.annotation.SuppressLint
import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.util.AttributeSet
import android.view.View
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.view.menu.ActionMenuItemView
import com.mikepenz.iconics.Iconics
import com.mikepenz.iconics.animation.tryToEnableIconicsAnimation
import com.mikepenz.iconics.core.R
import com.mikepenz.iconics.utils.buildIconics

internal object IconicsFactory {

    @JvmStatic
    fun onViewCreated(view: View?, context: Context, attrs: AttributeSet): View? {
        if (view != null && view.getTag(R.id.iconics_tag_id) != true) {
            onViewCreatedInternal(view, context, attrs)
            view.setTag(R.id.iconics_tag_id, true)
        }
        return view
    }

    @JvmStatic
    @SuppressLint("RestrictedApi")
    private fun onViewCreatedInternal(view: View, context: Context, attrs: AttributeSet?) {
        if (attrs == null) {
            return
        }

        when (view) {
            is ActionMenuItemView -> {
                IconicsAttrsApplier.getIconicsDrawable(context, attrs)?.let {
                    view.setIcon(view.tryToEnableIconicsAnimation(it))
                }
            }
            is EditText -> {
                // for an editText we only style initial as styling the Editable causes problems!
                view.buildIconics()
            }
            is TextView -> {
                view.buildIconics()

                view.addTextChangedListener(object : TextWatcher {
                    override fun beforeTextChanged(cs: CharSequence, i: Int, i1: Int, i2: Int) {
                    }

                    override fun onTextChanged(cs: CharSequence, i: Int, i1: Int, i2: Int) {
                    }

                    override fun afterTextChanged(editable: Editable) {
                        Iconics.styleEditable(editable)
                    }
                })
            }
            is ImageView -> {
                IconicsAttrsApplier.getIconicsDrawable(context, attrs)?.let {
                    view.setImageDrawable(view.tryToEnableIconicsAnimation(it))
                }
            }
        }
    }
}